### PR TITLE
core/bitarithm: introduce bitarithm_get/set_masked(), convert pca9685

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description
@gschorcht posted a collection of sensor drivers that all follow a common pattern. One neat helper function is `_set_reg_bit(uint8_t *byte, uint8_t mask, uint8_t bit)` that sets the masked bits in `byte` to the given value.

At first glance the function looks a bit odd as uses a loop to determine the lowest set bit of a compile-time constant. However, if `mask` is a compile-time constant, the compiler will [resolve the loop](https://godbolt.org/z/N_8UZD) and leave just a few instructions.

Now this pattern is quite useful for other drivers and to keep people from trying to optimize that loop that will be optimized away (for some reason, when replacing the loop with `bitarithm_lsb()`, the compiler will *not* optimize it away (only verified by increased binary size)), it's best to put it into common code. 

`bitarithm_set_masked()` is a macro only because this way it works for all types.

### Testing procedure

See if `tests/driver_pca9685` still produces the same binary.

### Issues/PRs references
#10083, #10082, #10092, #10420, #10462 all come with their own copy of this function.

### CODE REVIEW

File: drivers/pca9685/pca9685.c [hunk 1]
```diff
@@ -274,21 +273,21 @@ static int _init(pca9685_t *dev)
 
     /* set Auto-Increment flag */
     uint8_t byte = 0;
-    _set_reg_bit(&byte, PCA9685_MODE2_INVERT, dev->params.inv);
-    _set_reg_bit(&byte, PCA9685_MODE2_OUTDRV, dev->params.out_drv);
-    _set_reg_bit(&byte, PCA9685_MODE2_OUTNE, dev->params.out_ne);
+    bitarithm_set_masked(&byte, PCA9685_MODE2_INVERT, dev->params.inv);
```
Comment (gschorcht) [inline drivers/pca9685/pca9685.c]: I would prefer to make these changes within a separate PR once this PR is merged.

File: core/include/bitarithm.h [hunk 2]
```diff
@@ -163,6 +163,47 @@ static inline unsigned bitarithm_lsb(unsigned v)
...
+ * @note            The mask should be compile-time constant so the compiler
+ *                  is able to optimize the entire loop away.
+ *
+ * @param       val The value to modify.
+ * @param[in]  mask The bitmask to apply.
+ * @param[in]  bits The new value of the masked bits.
+ *
+ */
+#define bitarithm_set_masked(val, mask, bits) do {                  \
+        const unsigned _mask = mask;                                \
```
Comment (gschorcht) [inline core/include/bitarithm.h]: I would prefere a furthere indent for the code with exception of the closing `} while (0)` to see the structure of the macro a bit better.

Comment (gschorcht) [reply core/include/bitarithm.h]: Might there be a problem for `uint32_t` parameters on platforms where `unsigned` is just 16 bit?

File: drivers/pca9685/pca9685.c (hunk 1 above)
Comment (benpicco) [reply drivers/pca9685/pca9685.c]: Will do, just thought it's easier when I provide an example with it.

File: core/include/bitarithm.h (hunk 2 above)
Comment (benpicco) [reply core/include/bitarithm.h]: Yes - for the get I could get away with a macro to end up at a generic version, but I can't return from a macro so I was hoping `unsigned` will be enough.

If we can use C11 `_Generic` in RIOT, maybe this will be the solution here - I don't think any of our tool chains is *that* old that it would not support it.

I'll see what I can figure out - enjoy your hike!

Comment (kaspar030) [reply core/include/bitarithm.h]: IMO, if we're talking uint32_t and uint16_t, two implementations differing just in the type are preferable to a macro implementation. (That's also a coding convention - avoid macros if possible.)

**Comment by kaspar030** [comment]
> for some reason, when replacing the loop with `bitarithm_lsb()`, the compiler will _not_ optimize it away

That might be architecture dependent. ```bitarithm_lsb()``` can use multiple implementations (currently, the "naive" implementation using a loop, one using debruijn multiplication and __builtin_lsb()). On ARMs without CLZ, the debruijn method is the fastest in the scheduler use case, so that's used. That version pretty much prevents the compiler from recognizing any pattern... 

Could you try if the compiler optimizes correctly when using ```__builtin_lsb()```?

**Comment by kaspar030** [comment]
> One neat helper function is `_set_reg_bit(uint8_t *byte, uint8_t mask, uint8_t bit)` that sets the masked bits in `byte` to the given value.

An example would be nice. I don't quite get it. Like, byte=0b0110, mask=0b0011, bit=0b1001 => result = 0b0001"?

**Comment by kaspar030** [comment]
> Could you try if the compiler optimizes correctly when using `__builtin_ffs()`?

Seems like: https://godbolt.org/z/4U9iSn

File: core/include/bitarithm.h (hunk 2 above)
Comment (benpicco) [reply core/include/bitarithm.h]: uint8_t was the original use-case, so that makes it 3.

**Comment by benpicco** [comment]
> An example would be nice. I don't quite get it. Like, byte=0b0110, mask=0b0011, bit=0b1001 => result = 0b0001"?

Often times you have multiple variables in one register, e.g. bits 0-3 are `foo` and bits 4-7 are `bar`.

```C
union {
    uint8_t foo:4;
    uint8_t bar:4;
} byte
```

Now to set `bar` in `byte` you would do `mask=0b11110000`, the if `byte=0b00000110` and `bit=0b1001` => `result = 0b10010110`.

In your example `result` would be `0b0101`.

File: core/include/bitarithm.h (hunk 2 above)
Comment (kaspar030) [reply core/include/bitarithm.h]: When ```bitarithm_ffs()``` is used, they'd be one line each, right?

**Comment by kaspar030** [comment]
> Now to set `bar` in `byte` you would do `mask=0b11110000`, the if `byte=0b00000110` and `bit=0b1001` => `result = 0b10010110`.

Ah, thanks!

File: core/include/bitarithm.h (hunk 2 above)
Comment (kaspar030) [reply core/include/bitarithm.h]: Shouldn't this be equivalent?

```
static inline void bitarithm_set_masked_u32(uint32_t *val, uint32_t mask, uint32_t bits)
{
    *(val) = ((*(val) & ~mask) | ((bits << (__builtin_ffs(mask) - 1)) & mask));
}
```

(generated code looks different though)

Comment (gschorcht) [reply core/include/bitarithm.h]: Indeed, `__builtin_ffs` seems to produce exactly the same code on all our platforms. The long version `__builtin_ffsl(mask)` would have to be used for `uint32_t` since all our platforms define
```
#define __UINT32_TYPE__ long unsigned int
```

**Comment by fjmolinas** [comment]
@kaspar030 do you still have issues with this one?
